### ISSUE TITLE
Create a Candidate Extractor to match part numbers and storage temperature ranges

### ISSUE THREAD

**Issue by lukehsiao** [lukehsiao/HardwareDeepDive#13]
Now that we have naive implementations of #12 and #10, we should be able to move on to create the candidate extractor.

In the DeepDive tutorial, this would correspond to the ```has_spouse``` extractor that uses the data in the ```people_mentions``` table to determine whether or not they may be spouses. In our case, we will be using the ```product_num_mentions``` table and the ```storage_temp_mentions``` table to try to decide whether two entries are matched with each other.

**Comment by lukehsiao** [comment]
@xiao-cheng,

I'm getting ready to write the extractor for candidate relations. The spouse example tutorial says this:

> Now we need to create candidates for has_spouse relations. For simplicity, a relation candidate is a pair of person mentions in the same sentence.

> In order to train the system to decide whether a candidate is indeed expressing a marriage relation, we also need to generate some training data. However, it is hard to find ground truth on whether two mentions participate in has_spouse relation. Therefore, we use distant supervision rules that generate mention-level training data by heuristically mapping them to known entity-level relations in an existing knowledge base.

> The extractor we are now going to write takes all mentions of people in a sentence, and insert each pair of them into the table has_spouse, while also labeling some of these pairs as true or false according to the supervision rules.

In trying to draw a parallel to the example application, I believe we are set up to write an extractor that takes all mentions of product numbers and storage temperature ranges, while also labeling some of these pairs as true or false. Do you think this is a good approach? Any tips before I dig in?

**Comment by xiao-cheng** [comment]
Yes I think this is the right way. Don't worry about optimization too much right now unless it takes too long for a single file.

**Comment by lukehsiao** [comment]
The tutorial says this:
> For simplicity, a relation candidate is a pair of person mentions **in the same sentence**.
> We now define an extractor that creates all candidate relations and inserts them into the table has_spouse. We call them candidate relations because we do not know whether or not they are actually expressing a marriage relation: that's for DeepDive to predict later.

I will be mirroring this by defining a *candidate* as a part_num and storage_temp pair that appears **in the same document**. This should give us a large set of candidates without me filtering potential results. DeepDive can then do filtering based on *features*.

The ```part_temperature_candidates``` table will have a schema like this:

| Column           | Type    | Description                                                   |
|:-----------------|:--------|:--------------------------------------------------------------|
| doc_id           | text    | the document both are found in                                |
| part_pg_num      | bigint  | the page number this part num is on                           |
| part_table_num   | bigint  | the table number the part is found in                         |
| part_relation    | text    | the type of relation (header, row, etc) the part was found in |
| part_relation_id | bigint  | the id of the part relation                                   |
| part_num         | text    | the part number itself                                        |
| temp_pg_num      | bigint  | the page number the storage temp is on                        |
| temp_table_id    | bigint  | the table number the temp was found in                        |
| temp_row_id      | bigint  | the row id the temperature was found in                       |
| temp_min         | bigint  | the minimum storage temperature                               |
| temp_max         | bigint  | the maximum storage temperature                               |
| temp_unit        | text    | the units of the storage temperature                          |
| is_true          | boolean | whether this relation is correct                              |
| relation_id      | bigint  | unique identifier for part_temperature relationship           |
| id               | bigint  | reserved for DeepDive                                         |

*Note: I may need include column information for the storage temperatures later on, but for now, this should be ok.*

**Comment by lukehsiao** [comment]
It's part of this extractor (udf/ext_has_spouse.py) that the tutorial includes *distant supervision* using the ```spouses.csv``` and ```non-spouses.csv``` dictionaries to determine whether to guess true or false.

```py
# Load the spouse dictionary for distant supervision
spouses = defaultdict(lambda: None)
with open (APP_HOME + "/input/spouses.csv") as csvfile:
  reader = csv.reader(csvfile)
  for line in reader:
    spouses[line[0].strip().lower()] = line[1].strip().lower()

# Load relations of people that are not spouse
non_spouses = set()
lines = open(APP_HOME + '/input/non-spouses.tsv').readlines()
for line in lines:
  name1, name2, relation = line.strip().split('\t')
  non_spouses.add((name1, name2))  # Add a non-spouse relation pair

# For each input tuple
for row in sys.stdin:
  parts = row.strip().split('\t')
  if len(parts) != 5:
    print >>sys.stderr, 'Failed to parse row:', row
    continue

  sentence_id, p1_id, p1_text, p2_id, p2_text = parts

  p1_text = p1_text.strip()
  p2_text = p2_text.strip()
  p1_text_lower = p1_text.lower()
  p2_text_lower = p2_text.lower()

  # See if the combination of people is in our supervision dictionary
  # If so, set is_correct to true or false
  is_true = '\N'
  if spouses[p1_text_lower] == p2_text_lower:
    is_true = '1'
  if spouses[p2_text_lower] == p1_text_lower:
    is_true = '1'
  elif (p1_text == p2_text) or (p1_text in p2_text) or (p2_text in p1_text):
    is_true = '0'
  elif (p1_text_lower, p2_text_lower) in non_spouses:
    is_true = '0'
  elif (p2_text_lower, p1_text_lower) in non_spouses:
    is_true = '0'
```

I've implemented the code that pulls all the pairs of product numbers that are in the same document as storage temperatures. My plan is to use our hand-created gold data to do similar guessing for now.

However, I'm not sure how we will handle this when we want to run our 12,000 pdf dataset since we have no dictionaries for that huge set.

**Comment by lukehsiao** [comment]
Output currently looks like this:
```
lh@lh:~/repos/HardwareDeepDive$ deepdive sql "select * from part_temperature_candidates"
  doc_id   | part_pg_num | part_table_id | part_relation | part_relation_id | part_num | temp_pg_num | temp_table_id | temp_row_id | temp_min | temp_max | temp_unit |   relation_id   | is_true
-----------+-------------+---------------+---------------+------------------+----------+-------------+---------------+-------------+----------+----------+-----------+-----------------+---------
 bc550.pdf |           0 |             0 | header        |                0 | BC546    |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           0 |             0 | row           |                1 | BC546    |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           0 |             0 | row           |                2 | BC547    |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           0 |             0 | row           |                3 | BC548    |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           0 |             0 | row           |                4 | BC546    |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           0 |             0 | row           |                5 | BC547    |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           0 |             0 | row           |                6 | BC548    |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           0 |             0 | row           |                7 | BC546    |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           0 |             0 | row           |                8 | BC548    |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           0 |             1 | header        |                0 | BC546    |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           0 |             1 | row           |               12 | BC546    |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           0 |             1 | row           |               13 | BC549    |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           0 |             1 | row           |               14 | BC549    |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           0 |             1 | row           |               15 | 15000MHz |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           0 |             1 | row           |               15 | BC550    |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           0 |             1 | row           |                4 | 100mA    |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           0 |             1 | row           |                6 | 100mA    |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           0 |             1 | row           |                9 | 100MHz   |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           2 |             0 | header        |                2 | 27TYP    |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           2 |             0 | row           |               13 | 27TYP    |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           2 |             1 | header        |                2 | 27TYP    |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           3 |             0 | row           |                4 | Power247 |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
 bc550.pdf |           3 |             0 | row           |                9 | E2CMOS   |           0 |             0 |          12 |      -65 |      150 | C         | NOT_IMPLEMENTED | f
(23 rows)
```

**Comment by lukehsiao** [comment]
Feature Extractor is in place, although I only stuck in 2 features for now (same_row or same_page). Output looks like this:
```
             relation_id              |   feature
--------------------------------------+--------------
 bc550.pdf~^~100MHz~^~-65~^~150~^~C   | on_same_page
 bc550.pdf~^~BC549~^~-65~^~150~^~C    | on_same_page
 bc550.pdf~^~BC548~^~-65~^~150~^~C    | on_same_page
 bc550.pdf~^~BC546~^~-65~^~150~^~C    | on_same_row
 bc550.pdf~^~100mA~^~-65~^~150~^~C    | on_same_page
 bc550.pdf~^~BC546~^~-65~^~150~^~C    | on_same_page
 bc550.pdf~^~15000MHz~^~-65~^~150~^~C | on_same_page
 bc550.pdf~^~BC550~^~-65~^~150~^~C    | on_same_page
 bc550.pdf~^~BC547~^~-65~^~150~^~C    | on_same_page
(9 rows)
```

**Comment by lukehsiao** [comment]
We should refine this more later on. My calibration graphs look like this:
![part_temperature label](https://cloud.githubusercontent.com/assets/7573542/12188842/c72f8dda-b56d-11e5-8a9f-045bebcedb77.png)
when I run it on the 100 original PDFs.

**Comment by lukehsiao** [comment]
After fixing the labeling to have true, false, and '\N', our graph looks a lot better:
![part_temperature label](https://cloud.githubusercontent.com/assets/7573542/12373830/b1bca7fa-bc3c-11e5-86a7-0d8b44285dbe.png)

This is more what we expect. Strong trues because we have positive training data.

**Comment by ajratner** [comment]
Hey Luke,

Would love to chat about this in person next time we can meet up, but two
quick notes for now:

(a) From a high-level / quick glance, what you have here for relation
mention extraction looks like a good start

(b) Re: your question about distant supervision feasibility: most of our
traditional examples of distant supervision *rules* have been simple ones
that map from some tertiary gold set to the raw input... however, recently
our users (primarily in MEMEX, and me & others in genomics) have been
moving to using *any* rules / heuristics to programmatically supervise.
E.g. high precision patterns of any sort, not necessarily relying on some
external gold set

Now, we're in the process of building better tooling, formalisms and
algorithms for this... however, it's something to consider.  The key
high-level idea is that even right now, DeepDive should be able to take a
set of noisy / unreliable rules and over a large enough set of data come up
with a denoised result...

Anyway, can talk more in person, but just something to potentially try out

-Alex
